structwrapping/structwrapping.go

```diff
@@ -0,0 +1,154 @@
+package structwrapping
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+
+	wrapping "github.com/hashicorp/go-kms-wrapping"
+)
+
+type entry struct {
+	index int
+}
+
+type encDecMap map[string][2]*entry
+
+func buildEncDecMap(ctx context.Context, in interface{}) (encDecMap, error) {
+	val := reflect.ValueOf(in)
+	switch {
+	case !val.IsValid():
+		return nil, errors.New("input not valid")
+	case val.IsZero():
+		return nil, errors.New("input was not initialized")
+	case val.Kind() != reflect.Ptr:
+		return nil, errors.New("input not a pointer")
+	}
+
+	val = reflect.Indirect(val)
+	if val.Kind() != reflect.Struct {
+		return nil, errors.New("input not a struct")
+	}
+
+	typ := val.Type()
+	// plaintext,ciphertext
+	edMap := make(encDecMap, typ.NumField()/2)
+	for i := 0; i < typ.NumField(); i++ {
+		field := typ.Field(i)
+		tag, ok := field.Tag.Lookup("wrapping")
+		if !ok {
+			continue
+		}
+		tagParts := strings.Split(tag, ",")
+		if len(tagParts) != 2 {
+			return nil, errors.New("error in wrapping tag specification")
+		}
+
+		fieldKind := field.Type.Kind()
+		switch tagParts[0] {
+		case "pt":
+			if fieldKind != reflect.Slice {
+				return nil, errors.New("plaintext value is not a slice")
+			}
+			if !field.Type.ConvertibleTo(reflect.TypeOf([]byte(nil))) {
+				return nil, errors.New("plaintext value is not a byte slice")
+			}
+			curr := edMap[tagParts[1]]
+			if curr[0] != nil {
+				return nil, errors.New("detected two pt wrapping tags with the same identifier")
+			}
+			curr[0] = &entry{index: i}
+			edMap[tagParts[1]] = curr
+
+		case "ct":
+			if fieldKind != reflect.Ptr {
+				return nil, errors.New("ciphertext value is not a pointer")
+			}
+			if !field.Type.ConvertibleTo(reflect.TypeOf((*wrapping.EncryptedBlobInfo)(nil))) {
+				return nil, errors.New("ciphertext value is not the expected type")
+			}
+			curr := edMap[tagParts[1]]
+			if curr[1] != nil {
+				return nil, errors.New("detected two ct wrapping tags with the same identifier")
+			}
+			curr[1] = &entry{index: i}
+			edMap[tagParts[1]] = curr
+
+		default:
+			return nil, errors.New("unknown tag type for wrapping tag")
+		}
+	}
+
+	for k, v := range edMap {
+		if v[0] == nil {
+			return nil, fmt.Errorf("no pt wrapping tag found for identifier %q", k)
+		}
+		if v[1] == nil {
+			return nil, fmt.Errorf("no ct wrapping tag found for identifier %q", k)
+		}
+	}
+
+	return edMap, nil
+}
+
+func WrapStruct(ctx context.Context, wrapper wrapping.Wrapper, in interface{}) error {
+	if wrapper == nil {
```

Comment:
perhaps check ctx for nil?  if you ever decide to use the ctx in the future, you could have a lot broken clients that have been passing nil

Reply:
To be honest, I'm not sure how libraries writ large, by which I mean Go stdlib and others, handle nil ctxes. Considering it's a pure passthrough though I don't really feel like this is the place to take a stand.